runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -84,11 +84,11 @@ protected ILogData removeFromQueue(NavigableSet<Long> queue) {
                 // to be serviced immediately, rather than reading one entry at a time.
                 ld = read(currentRead, queue);
 
-                if (queue == getCurrentContext().readQueue && ld != null) {
+                if (queue == getCurrentContext().readQueue) {
                     // Validate that the data entry belongs to this stream, otherwise, skip.
                     // This verification protects from sequencer regression (tokens assigned in an older epoch
                     // that were never written to, and reassigned on a newer epoch)
-                    if (ld.containsStream(this.id) && ld.getType() == DataType.DATA) {
+                    if (ld.isHole() || (ld.containsStream(this.id) && ld.isData())) {
```

Comment:
I feel this might not be completely correct in case of sequencer regress.  I think we need to check containsStream() even for a hole, as a hole could be a real hole, which has no stream info, or a no-op which still belongs to some stream. Consider this case:
1) In epoch 0, sequencer issues a token with address 100 to stream A.
2) No one writes to this token, but the client queries sequencer for stream A and get back a bitmap that 100 belongs to stream A.
3) Epoch changes to 1, new sequencer boots up, recovers the stream tail map from log unit, since no one writes to that token with address 100, log unit does not know about that token, neither does new sequencer.
4) New sequencer issues token 100 to stream B to new client 
5) New client uses this token to enforce a hole for stream B (the checkpoint hole).
6) Now the old client which operates on stream A finish the read, but it reads a hole with backpointer of stream B, since it's a hole, you will update the stream A's pointer to point to 100, which belongs to stream B and also put it in to the resolved queue.

So in this case we end up pointing stream A's pointer to a hole that belongs to stream B and put it into A's resolved queue.

Reply:
Yes, I added check containsStream() for a no-op hole